### PR TITLE
Support a GraphML document which has multiple graphs.

### LINKED ISSUE #6: Empty edge is allowed by GraphML spec but graphml-js can't parse a graph with no edge.
graphml-js can't parse a graph which has no edge and reports the following error:

```bash
$ cat index.js
var graphml = require('graphml-js')
  , fs      = require('fs');

var content = fs.readFileSync('./graph.xml');
var parser  = new graphml.GraphMLParser();

parser.parse(content, function(error, graph) {
  console.log(graph);
});

$ cat graph.xml
<graphml xsi:schemaLocation="http://graphml.graphdrawing.org/xmlns http://graphml.graphdrawing.org/xmlns/1.0/graphml.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns="http://graphml.graphdrawing.org/xmlns">                               
    <graph id="G" edgedefault="directed">                                       
        <node id="1"></node>
    </graph>                            
</graphml>

$ node index.js
events.js:182
      throw er; // Unhandled 'error' event
      ^

TypeError: Cannot read property 'length' of undefined
    at GraphMLParser.buildEdges (/Users/syamaoka/tmp/graphmltest/node_modules/graphml-js/dist/graphml.js:129:56)
    at /Users/syamaoka/tmp/graphmltest/node_modules/graphml-js/dist/graphml.js:101:19
    at Parser.<anonymous> (/Users/syamaoka/tmp/graphmltest/node_modules/xml2js/lib/parser.js:303:18)
    at emitOne (events.js:115:13)
    at Parser.emit (events.js:210:7)
    at SAXParser.onclosetag (/Users/syamaoka/tmp/graphmltest/node_modules/xml2js/lib/parser.js:261:26)
    at emit (/Users/syamaoka/tmp/graphmltest/node_modules/sax/lib/sax.js:624:35)
    at emitNode (/Users/syamaoka/tmp/graphmltest/node_modules/sax/lib/sax.js:629:5)
    at closeTag (/Users/syamaoka/tmp/graphmltest/node_modules/sax/lib/sax.js:889:7)
    at SAXParser.write (/Users/syamaoka/tmp/graphmltest/node_modules/sax/lib/sax.js:1436:13)
```

The GraphML specification defines a graph as follows (excerpt from http://graphml.graphdrawing.org/xmlns/1.1/graphml-structure.xsd):

```xml
<xs:complexType name="graph.type" final="#all">
    <xs:annotation>
        <xs:documentation source="http://graphml.graphdrawing.org/" xml:lang="en">Complex type for the <graph> element.</xs:documentation>
    </xs:annotation>
    <xs:sequence>
        <xs:element ref="desc" minOccurs="0"/>
        <xs:choice>
            <xs:sequence>
                <xs:choice minOccurs="0" maxOccurs="unbounded">
                    <xs:element ref="data"/>
                    <xs:element ref="node"/>
                    <xs:element ref="edge"/>
                    <xs:element ref="hyperedge"/>
                </xs:choice>
            </xs:sequence>
            <xs:element ref="locator"/>
        </xs:choice>
    </xs:sequence>
```

This XMLSchema defines the elements of `graph.type` as `minOccurs=0`, so it means that a graph which has no data, node, edge, and hyperedge is allowed.

### PR DESCRIPTION
**Pull request by shinnya**
This PR resolves #6 and #7 .

This change is not compatible with the previous version of GraphMLParser. However, it might not matter because the previous implementation doesn't satisfy the GraphML specification anyway.

# Changes
* Added `src/schema.ts` for code completion.
* Fixed a bug that GraphMLParser can't parse a GraphML document which has no edge. Nodes of a result graph after parsing is an empty array when no edge is defined in a GraphML document.
* Changed the API of GraphMLParser.parse. The callback function passed to GraphML.parse receives an array of Graph as second argument.

Best regards.

### CODE REVIEW

File: src/parser.ts [hunk 1]
```diff
@@ -90,6 +91,24 @@ interface AttributeKeyMap {
...
+            }
+
+            let graphs = document.graphml.graph;
+            let keys = document.graphml.key;
+            cb(err, graphs.map((graph: schema.GraphElement) => new GraphBuilder().build(keys, graph)));
+        });
+    }
+}
+
+class GraphBuilder {
```
Comment (shinnya) [inline src/parser.ts]: I renamed `GraphMLParser` to `GraphBuilder` and this class is used by new `GraphMLParser`. The reason of this change is that we should parse a graph multiple times.

File: src/parser.ts [hunk 2]
```diff
@@ -90,6 +91,24 @@ interface AttributeKeyMap {
 }
 
 export class GraphMLParser {
+    public parse(text: string, cb?: Function) {
+
+        let parser = new Parser();
+
+        parser.parseString(text, (err: any, document: schema.GraphMLDocument) => {
```
Comment (shinnya) [inline src/parser.ts]: Now receives the result of parsing as a `schema.GraphMLDocument` for improving code completion.

**Comment by shinnya** [comment]
@anderspitman  BTW do you have any plan to release a new version in a few days? I want you to include this change into the new version.

File: src/parser.ts (hunk 1 above)
Comment (anderspitman) [reply src/parser.ts]: This all looks good! The only thing is that generally the putting "Builder" on the end of a name suggests that the builder pattern is being used, which isn't  really the case here. Is have a build method really necessary anymore? Maybe we should rename `GraphBuilder` to something like `Graph` or `ParsedGraph` and pass the keys and graph directly to the constructor?

**Comment by anderspitman** [comment]
You mean on npm? I'm happy to push a new version whenever you like. I'm not currently using it anymore so breaking compatibility isn't a problem for me at all.

**Comment by shinnya** [comment]
> You mean on npm?

Yes, npm.

> I'm happy to push a new version whenever you like.

Thank you for your kindness. Could you please release a new version after merging this PR 😄 

> I'm not currently using it anymore so breaking compatibility isn't a problem for me at all.

OK.

File: src/parser.ts (hunk 1 above)
Comment (shinnya) [reply src/parser.ts]: Good point. 

There are some options like:

1. remove `GraphBuilder` and move all functions to `Graph` class. Then add `Graph.create(GraphKeyElement[], GraphElement)` static function, which is a factory method for `Graph`.
2. rename `GraphBuilder` to `ParsedGraph` and pass the keys and graph directly to the constructor (you mentioned above). 

I think 1 is simple enough in this case.

**Comment by shinnya** [comment]
oops. There is something wrong with the change.

```
(1,24): error TS2307: Cannot find module './schema'.
```

[2017/10/17 16:16:00 +09:00] fixed https://github.com/anderspitman/graphml-js/pull/8/commits/a278d6e317dd41956b7d1d441bce8587677b22b3

File: dist/index.d.ts [hunk 3]
```diff
@@ -1,35 +1 @@
...
-    private keys;
-    private graph;
-    constructor();
-    parse(text: string, cb?: Function): void;
-    private buildKeys(data);
-    private buildNodes(data);
-    private buildEdges(data);
-    private buildAttributes(newAttr, attributes);
-}
+export * from './parser';
```
Comment (shinnya) [inline dist/index.d.ts:1]: Added `index.d.ts` to make it easier to aggregate all d.ts files in gulpfile.js. See gulptifle.js.

**Comment by shinnya** [comment]
Could you review my change again?

**Comment by anderspitman** [comment]
LGTM

**Comment by anderspitman** [comment]
Thanks!

**Pull request merged**